### ISSUE TITLE
Missing mean_horizontal and more

### ISSUE THREAD

**Issue by thomasaarholt** [pola-rs/polars#9966]
### Checks
- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description
With the latest version of polars, I am not able to find a way to do a horizontal mean without knowing how many columns I am aggregating! `pl.mean(["col1", "col2"])` doesn't run, and `pl.col(["col1", "col2"]).mean()` returns two columns with their means.

The mean and median [were requested in discord](https://discord.com/channels/908022250106667068/1014967651656814694/1131131421860302918).

We're currently missing:
- `mean_horizontal`
- `median_horizontal`
- `std_horizontal`
- `var_horizontal`
And maybe also:
- `count_horizontal` (Might be useful where you want to do some other kind of averaging, and need to divide by the number of columns in the expression?)

How I am doing horizontal mean for now:
```python
import polars as pl
df = pl.DataFrame({"col1":[1,2,3], "col2":[4,5,6]})

num_columns = 2
df.with_columns(pl.sum_horizontal(["col1", "col2"]) / num_columns)
```

**Comment by stinodego** [comment]
Just to clarify, the reason that these are missing is that they are not a simple wrapper for fold. There is some additional logic required.

This is not a bug. I'd say you can add your feature requests to #9963 for now. I'll close this as a duplicate.

**Comment by stinodego** [comment]
On second thought, I will make this the leading issue on this topic, as it is more complete.

**Comment by moritzwilksch** [comment]
In the same vein, `.arg_max_horizontal()` and `.arg_min_horizontal()` would be great additions, too!

**Comment by stfdxv** [comment]
To count non-null values horizontally instead of  `count_horizontal` (which is not implemented yet) I use:
```python
df.with_columns(pl.sum_horizontal(pl.col(["col1", "col2"]).is_not_null()).alias("count"))
```

**Comment by Wainberg** [comment]
`.cummax()`, `.cummin()`, `.cumprod()`, `.n_unique()`, `.product()`, and `.quantile()` are also missing horizontal versions.

**Comment by KDruzhkin** [comment]
While this is still up in the air...

It might be good to group “horizontal_” methods somehow, to make them more discoverable.

Renaming to “horizontal_mean”, “horizontal_median” etc. Or putting them in a separate namespace: “horizontal.mean” etc.

**Comment by stinodego** [comment]
> It might be good to group “horizontal_” methods somehow, to make them more discoverable.
> 
> Renaming to “horizontal_mean”, “horizontal_median” etc. Or putting them in a separate namespace: “horizontal.mean” etc.

This has been brought up before, but we currently don't see a way to do it which is a clear improvement over the status quo.

**Comment by mincej** [comment]
> How I am doing horizontal mean for now:

A way that I've worked around this for more complicated instances, such as median, is:
1. Concatenate columns you want into list with `concat_list` (assumes each cell is not an array-like)
2. Using `map_elements` to apply a function to work on the list in each row
3. Saving result with new alias 

The logic being that a temporary representation of the values of the needed columns is created as a list, which can then be operated on by third-party functions as a single array-like. 

Here's an example with `np.median`: 
```
dataframe = pl.DataFrame(data = [[0, 1, 2], [3, 4, 5], [6, 7, 8]], schema = ["a", "b", "c"])

dataframe = dataframe.with_columns(

    # Median of all columns
    (
        pl.concat_list(pl.all())
        .map_elements(lambda l: np.median(l), return_dtype = pl.Float64)
        .alias("median_all")
    ),

    # Median of all columns except 'a'
    (
        pl.concat_list(pl.all().exclude("a"))
        .map_elements(lambda l: np.median(l), return_dtype = pl.Float64)
        .alias("median_noa")
    )
)

print(dataframe)
shape: (3, 5)

┌─────┬─────┬─────┬────────────┬────────────┐
│ a   ┆ b   ┆ c   ┆ median_all ┆ median_noa │
│ --- ┆ --- ┆ --- ┆ ---        ┆ ---        │
│ i64 ┆ i64 ┆ i64 ┆ f64        ┆ f64        │
╞═════╪═════╪═════╪════════════╪════════════╡
│ 0   ┆ 3   ┆ 6   ┆ 3.0        ┆ 4.5        │
│ 1   ┆ 4   ┆ 7   ┆ 4.0        ┆ 5.5        │
│ 2   ┆ 5   ┆ 8   ┆ 5.0        ┆ 6.5        │
└─────┴─────┴─────┴────────────┴────────────┘
```

**Comment by deanm0000** [comment]
In general, wouldn't it be cleaner to implement more aggregations in `.list` and `.arr` then do `concat_list('col1','col2'...).list.var()` That way there doesn't need to be a separate horizontal and a list implementation of all these functions or would the horizontal implementation just be a syntax sugar way of doing the `concat_list` approach?

@mincej For your example, you don't need `map_elements` you can do `explode().any_pl_op_you_want().over(pl.int_range(pl.len()))`

For instance:
```
df.with_columns(
    pl.concat_list(pl.all())
    .explode().median().over(pl.int_range(pl.len()))
    .alias('median_all')
    )
```

**Comment by mincej** [comment]
This is also a totally valid approach, but the reason that I use `map_elements` is to account for functions that may not be built into polars, or if more data is needed to run the function - for example, I'm using this approach to interpolate one column based on another column using `np.interp` while the Polars team works on supporting interpolating by a pl.Float64 column.

**Comment by mincej** [comment]
Thanks for the pointers - `map_elements` made sense to me logically, but learning `pl.len()`'s context and working with `pl.int_range()` will come in handy in the future!

**Comment by y9c** [comment]
Will this API become more consistent by refactoring all instances of `pl.func_horizontal(...)` to a structure like `pl.horizontal(...).func()`?

**Comment by rhug123** [comment]
I agree, currently we just have horizontal functions for `all`, `any`, `max`, `min`, `sum`, `mean` and `cum_sum`

I think adding horizontal versions for `cum_max`, `cum_min`, `cum_prod` would be helpful

having `horizontal_forward_fill`, `horizontal_backward_fill` would be helpful as well

More generally, what is the philosophy behind which Polars functions receive horizontal variants? Is the long-term goal for most functions to eventually gain horizontal equivalents where appropriate?
